File: src/main/java/br/com/tech/challenge/service/AssociateService.java
package br.com.tech.challenge.service;


import br.com.tech.challenge.cpf.CPFValidator;
import br.com.tech.challenge.exceptions.InvalidDocumentException;
import br.com.tech.challenge.model.Associate;
import br.com.tech.challenge.repository.AssociateRepository;
import lombok.AllArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@AllArgsConstructor
public class AssociateService {
    private final AssociateRepository associateRepository;
    private final CPFValidator validator;

    public Associate findAssociateByCpfOrCreateNewAssociate(String cpf) {
        return associateRepository.findByCpf(cpf).orElseGet(() -> createNewAssociate(new Associate(cpf)));
    }

    public Associate createNewAssociate(Associate associate) {
        if (!cpfIsValid(associate.getCpf())) {
            throw new InvalidDocumentException(STR."Document {CPF} = \{associate.getCpf()} is not valid");
        }
        return associateRepository.save(associate);
    }

    public boolean cpfIsValid(String cpf) {
        return validator.test(cpf).isValid();
    }
}

File: src/main/java/br/com/tech/challenge/service/AssociateService.kt
package br.com.ntconsult.challenge.service

import br.com.ntconsult.challenge.exception.InvalidDocumentException
import br.com.ntconsult.challenge.model.Associate
import br.com.ntconsult.challenge.proxy.cpf.CPFValidator
import br.com.ntconsult.challenge.repository.AssociateRepository
import org.springframework.stereotype.Service

@Service
class AssociateService(private val associateRepository: AssociateRepository,
                       private val validator: CPFValidator) {

    fun findAssociateByCpfOrCreateNewAssociate(cpf: String): Associate {
        return associateRepository.findByCpf(cpf).orElseGet {
            createNewAssociate(Associate(cpf = cpf))
        }
    }

    fun createNewAssociate(associate: Associate): Associate {
        if(!cpfIsValid(associate.cpf)) {
            throw InvalidDocumentException("Document {CPF} = ${associate.cpf} is not valid")
        }
        return associateRepository.save(associate)
    }

    fun cpfIsValid(cpf: String): Boolean {
        return validator.cpfIsValid(cpf).isValid()
    }

}
